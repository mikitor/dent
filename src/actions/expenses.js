import uuid from 'uuid';

const addExpense = ({ description = '', note = '', amount = 0, createdAt } = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    note,
    amount,
    createdAt
  }
});

const deleteExpense = ({ id } = {}) => ({
  type: 'DELETE_EXPENSE',
  id
});

const editExpense = (id, update) => ({
  type: 'EDIT_EXPENSE',
  id,
  update
});

export { addExpense, deleteExpense, editExpense };
