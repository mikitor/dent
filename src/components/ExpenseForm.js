import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      focused: false,
      error: ''
    };
  }

  handleChange = e => {
    const { value, name } = e.target;
    this.setState({ [name]: value });
  };

  handleDateChange = createdAt => {
    if (createdAt) {
      this.setState({ createdAt });
    }
  };

  handleAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState({ amount });
    }
  };

  handleSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState({ error: 'Please provide description and amount' });
    } else {
      this.setState({ error: '' });
      const { description, amount, note, createdAt } = this.state;
      this.props.onSubmit({
        description,
        note,
        amount: parseFloat(amount) * 100,
        createdAt: moment(createdAt).valueOf()
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.handleSubmit}>
          <h4>Expenseform</h4>
          <label htmlFor="description">
            Description
            <input
              type="text"
              name="description"
              value={this.state.description}
              onChange={this.handleChange}
              placeholder="Description"
              autoFocus
            />
          </label>
          <label htmlFor="note">
            Note
            <textarea
              name="note"
              placeholder="Note"
              value={this.state.note}
              onChange={this.handleChange}
            />
          </label>
          <label htmlFor="amount">
            Amount
            <input
              type="text"
              name="amount"
              placeholder="Amount"
              value={this.state.amount}
              onChange={this.handleAmountChange}
            />
          </label>
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.handleDateChange}
            focused={this.state.focused}
            onFocusChange={({ focused }) => this.setState({ focused })}
            id="createdAt"
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <input type="submit" value="Add expense" />
        </form>
      </div>
    );
  }
}
