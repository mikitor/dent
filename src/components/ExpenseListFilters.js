import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      focusedInput: null
    };
  }

  handleTextChange = e => {
    this.props.setTextFilter(e.target.value);
  };

  handleSortByChange = e => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  handleDatesChange = ({ startDate, endDate }) => {
    this.props.setEndDate(endDate);
    this.props.setStartDate(startDate);
  };

  handleFocusChange = focusedInput => this.setState({ focusedInput });

  render() {
    return (
      <div>
        <input type="text" value={this.props.filters.text} onChange={this.handleTextChange} />
        <select value={this.props.filters.sortBy} onChange={this.handleSortByChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        {this.props.filters.sortBy === 'date' && (
          <DateRangePicker
            startDate={this.props.filters.startDate}
            startDateId="startDate"
            endDate={this.props.filters.endDate}
            endDateId="endDate"
            onDatesChange={this.handleDatesChange}
            focusedInput={this.state.focusedInput}
            onFocusChange={this.handleFocusChange}
            isOutsideRange={() => false}
            numberOfMonths={1}
            showClearDates
          />
        )}

        <button type="submit">Filter</button>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
  setStartDate: startDate => dispatch(setStartDate(startDate))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExpenseListFilters);
