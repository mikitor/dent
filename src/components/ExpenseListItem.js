import React from 'react';
import moment from 'moment';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
  <div>
    <Link to={`/edit/${id}`}>{description}</Link>
    <p>Amount: {amount}</p>
    <p>CreatedAt: {moment(createdAt).format('LL')}</p>
  </div>
);

export default ExpenseListItem;
