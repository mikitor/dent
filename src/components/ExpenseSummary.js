import React from 'react';
import { connect } from 'react-redux';
import getTotalExpenses from '../utils/getTotalExpenses';
import getVisibleExpenses from '../utils/getVisibleExpenses';

export const ExpenseSummary = ({ expenseCount, expensesTotal }) => {
  const pluralize = expenseCount === 1 ? 'expense' : 'expenses';
  return (
    <div>
      <h1>
        Viewing {expenseCount} {pluralize} totalling ${expensesTotal / 100}
      </h1>
    </div>
  );
};

const mapStateToProps = state => {
  const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: getTotalExpenses(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpenseSummary);
