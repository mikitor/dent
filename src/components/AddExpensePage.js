import React from 'react';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
  handleSubmit = expense => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <h1>Add expense</h1>
        <ExpenseForm onSubmit={this.handleSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(addExpense(expense))
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
