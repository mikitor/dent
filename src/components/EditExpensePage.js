import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, deleteExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  handleSubmit = expense => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  handleDelete = e => {
    this.props.deleteExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm onSubmit={this.handleSubmit} expense={this.props.expense} />
        <button type="button" onClick={this.handleDelete}>
          Delete
        </button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  deleteExpense: id => dispatch(deleteExpense(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
