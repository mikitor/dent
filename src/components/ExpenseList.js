import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../utils/getVisibleExpenses';

export const ExpenseList = props => (
  <div>
    {props.expenses.length === 0 ? (
      <p>No expenses</p>
    ) : (
      <ul>
        {props.expenses.map((expense, index) => (
          <ExpenseListItem {...expense} key={index} />
        ))}
      </ul>
    )}
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
