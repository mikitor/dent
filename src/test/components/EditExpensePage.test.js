import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import { expenseOne, expenseTwo } from '../fixtures/expenses';

let editExpense, deleteExpense, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  deleteExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      expense={expenseOne}
      deleteExpense={deleteExpense}
      editExpense={editExpense}
      history={history}
    />
  );
});

it('should render EditExpensePage', () => {
  expect(wrapper).toMatchSnapshot();
});

it('should call delete expense with valid props', () => {
  wrapper.find('button').simulate('click');
  expect(deleteExpense).toHaveBeenCalledWith({ id: expenseOne.id });
  expect(history.push).toHaveBeenCalledWith('/');
});

it('should call edit expense with valid props', () => {
  const { note, description, amount, createdAt } = expenseTwo;
  wrapper.find('ExpenseForm').simulate('submit', { note, description, amount, createdAt });
  expect(editExpense).toHaveBeenCalledWith(expenseOne.id, { note, description, amount, createdAt });
  expect(history.push).toHaveBeenCalledWith('/');
});
