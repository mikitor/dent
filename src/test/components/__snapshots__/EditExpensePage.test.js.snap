// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 400,
        "createdAt": 14,
        "description": "Test one desc",
        "id": "124",
        "note": "Test one note",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
    type="button"
  >
    Delete
  </button>
</div>
`;
